server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -1525,19 +1529,40 @@ private CanMatchResponse canMatch(ShardSearchRequest request, boolean checkRefre
                 final boolean aliasFilterCanMatch = request.getAliasFilter().getQueryBuilder() instanceof MatchNoneQueryBuilder == false;
                 FieldSortBuilder sortBuilder = FieldSortBuilder.getPrimaryFieldSortOrNull(request.source());
                 MinAndMax<?> minMax = sortBuilder != null ? FieldSortBuilder.getMinMaxOrNull(context, sortBuilder) : null;
-                final boolean canMatch;
+                boolean canMatch;
                 if (canRewriteToMatchNone(request.source())) {
                     QueryBuilder queryBuilder = request.source().query();
                     canMatch = aliasFilterCanMatch && queryBuilder instanceof MatchNoneQueryBuilder == false;
+                    canMatch = aliasFilterCanMatch && queryBuilder instanceof MatchNoneQueryBuilder == false;
                 } else {
                     // null query means match_all
                     canMatch = aliasFilterCanMatch;
                 }
+                canMatch = canMatch && canMatchSearchAfter(searchContext, minMax);
+
                 return new CanMatchResponse(canMatch || hasRefreshPending, minMax);
             }
         }
     }
 
+    public static boolean canMatchSearchAfter(SearchContext searchContext, MinAndMax<?> minMax) {
+        if (searchContext != null && searchContext.searchAfter() != null && minMax != null) {
+            SortOrder order = searchContext.request().source().sorts().get(0).order();
+            if (order == SortOrder.DESC) {
+                if (minMax.compareMin(searchContext.searchAfter().fields[0]) > 0) {
```

Comment:
Minor, we could extract the `searchContext.searchAfter().fields[0]` to local variable